src/monitor/node_active_protocol.c

```diff
@@ -1259,6 +1260,120 @@ perform_failover(PG_FUNCTION_ARGS)
 }
 
 
+/*
+ * promote promotes a given target node in a group.
+ */
+Datum
+perform_promotion(PG_FUNCTION_ARGS)
+{
+	text *formationIdText = PG_GETARG_TEXT_P(0);
+	char *formationId = text_to_cstring(formationIdText);
+
+	text *nodeNameText = PG_GETARG_TEXT_P(1);
+	char *nodeName = text_to_cstring(nodeNameText);
+
+	List *groupNodesList = NULL;
+	int totalNodesCount = 0;
+
+	AutoFailoverNode *currentNode = NULL;
+
+	checkPgAutoFailoverVersion();
+
+	currentNode = GetAutoFailoverNodeByName(formationId, nodeName);
+
+	if (currentNode == NULL)
+	{
+		ereport(ERROR,
+				(errmsg("node \"%s\" is not registered in formation \"%s\"",
+						nodeName, formationId)));
+	}
+
+	LockFormation(formationId, ShareLock);
+	LockNodeGroup(formationId, currentNode->groupId, ExclusiveLock);
+
+	/*
+	 * If the current node is the primary, that's done.
+	 */
+	if (IsCurrentState(currentNode, REPLICATION_STATE_SINGLE) ||
+		IsCurrentState(currentNode, REPLICATION_STATE_PRIMARY))
+	{
+		/* return false: no promotion is happening */
+		PG_RETURN_BOOL(false);
+	}
+
+	/*
+	 * If we have only two nodes in the group, then perform a failover.
+	 */
+	groupNodesList =
+		AutoFailoverNodeGroup(currentNode->formationId, currentNode->groupId);
+
+	totalNodesCount = list_length(groupNodesList);
+
+	if (totalNodesCount <= 2)
+	{
+		DirectFunctionCall2(perform_failover,
+							CStringGetTextDatum(formationId),
+							Int32GetDatum(currentNode->groupId));
+
+		/* if we reach this point, then a failover is in progress */
+		PG_RETURN_BOOL(true);
+	}
+	else
+	{
+		char message[BUFSIZE] = { 0 };
+
+		/*
+		 * In the general case, we perform a little trick:
+		 *
+		 *  - first increment the node's candidate-priority by 100,
+		 *
+		 *  - then call perform_failover,
+		 *
+		 *  - when the node reaches WAIT_PRIMARY again, after promotion, reset
+		 *    its candidate priority.
+		 */
+		if (currentNode->candidatePriority == 0)
+		{
+			ereport(ERROR,
+					(errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
+					 errmsg("cannot perform promotion: node %s in formation %s "
+							"has a candidate priority of 0 (zero).",
+							nodeName, formationId)));
+		}
+
+		currentNode->candidatePriority += 100;
```

Comment:
minor: Maybe use a macro instead of `100` for _readability,_ such as `MAX_CANDIDATE_PRIORITY`. But, if you  decide to do that, we  should also go over the uses of `100`  in the code  and use  the macro for completeness.

Reply:
Good point, now added with a comment that also covers the user range being 0..100 while the internal range is different (0..200).